### PR TITLE
SriovNetwork: react on namespace creation

### PR DESCRIPTION
**Pull request by zeeke**
If a user creates an SriovNetwork with a network namespace that doesn't exist, retrying reconciling with an exponential backoff is not efficient, as the routing will fail until the namespace is created.

This patch makes the controller watch Namespace resource creation and reconcile the related SriovNetworks if needed.

Before this PR, the SriovNetwork reconcile function returned an error, triggering the exponential back-off retry system:

```
2023-08-11T16:14:18.839626394Z  ERROR   Couldn't create NetworkAttachmentDefinition CR  {"controller": "sriovnetwork", "controllerGroup": "sriovnetwork.openshift.io", "controllerKind": "SriovNetwork", "SriovNetwork": {"name":"test-missing-ns","namespace":"openshift-sriov-network-operator"}, "namespace": "openshift-sriov-network-operator", "name": "test-missing-ns", "reconcileID": "9783b6b4-b193-4d78-9371-be4f3487d68b", "sriovnetwork": "openshift-sriov-network-operator/test-missing-ns", "Namespace": "ns-xxx", "Name": "test-missing-ns", "error": "namespaces \"ns-xxx\" not found"}
2023-08-11T16:14:18.905244077Z  ERROR   Couldn't create NetworkAttachmentDefinition CR  {"controller": "sriovnetwork", "controllerGroup": "sriovnetwork.openshift.io", "controllerKind": "SriovNetwork", "SriovNetwork": {"name":"test-missing-ns","namespace":"openshift-sriov-network-operator"}, "namespace": "openshift-sriov-network-operator", "name": "test-missing-ns", "reconcileID": "fccf9f74-f963-47f4-b672-d5099eb40484", "sriovnetwork": "openshift-sriov-network-operator/test-missing-ns", "Namespace": "ns-xxx", "Name": "test-missing-ns", "error": "namespaces \"ns-xxx\" not found"}
2023-08-11T16:14:18.970988294Z  ERROR   Couldn't create NetworkAttachmentDefinition CR  {"controller": "sriovnetwork", "controllerGroup": "sriovnetwork.openshift.io", "controllerKind": "SriovNetwork", "SriovNetwork": {"name":"test-missing-ns","namespace":"openshift-sriov-network-operator"}, "namespace": "openshift-sriov-network-operator", "name": "test-missing-ns", "reconcileID": "5c790fc5-ca13-4283-b0a8-22753d0ccb89", "sriovnetwork": "openshift-sriov-network-operator/test-missing-ns", "Namespace": "ns-xxx", "Name": "test-missing-ns", "error": "namespaces \"ns-xxx\" not found"}
2023-08-11T16:14:19.047349515Z  ERROR   Couldn't create NetworkAttachmentDefinition CR  {"controller": "sriovnetwork", "controllerGroup": "sriovnetwork.openshift.io", "controllerKind": "SriovNetwork", "SriovNetwork": {"name":"test-missing-ns","namespace":"openshift-sriov-network-operator"}, "namespace": "openshift-sriov-network-operator", "name": "test-missing-ns", "reconcileID": "388e00a0-dfe3-4b01-afef-b2ec0b62a675", "sriovnetwork": "openshift-sriov-network-operator/test-missing-ns", "Namespace": "ns-xxx", "Name": "test-missing-ns", "error": "namespaces \"ns-xxx\" not found"}
2023-08-11T16:14:19.142652361Z  ERROR   Couldn't create NetworkAttachmentDefinition CR  {"controller": "sriovnetwork", "controllerGroup": "sriovnetwork.openshift.io", "controllerKind": "SriovNetwork", "SriovNetwork": {"name":"test-missing-ns","namespace":"openshift-sriov-network-operator"}, "namespace": "openshift-sriov-network-operator", "name": "test-missing-ns", "reconcileID": "dd8a06d4-08d8-479f-b5c4-b29e2d07a6b9", "sriovnetwork": "openshift-sriov-network-operator/test-missing-ns", "Namespace": "ns-xxx", "Name": "test-missing-ns", "error": "namespaces \"ns-xxx\" not found"}
2023-08-11T16:14:19.278729554Z  ERROR   Couldn't create NetworkAttachmentDefinition CR  {"controller": "sriovnetwork", "controllerGroup": "sriovnetwork.openshift.io", "controllerKind": "SriovNetwork", "SriovNetwork": {"name":"test-missing-ns","namespace":"openshift-sriov-network-operator"}, "namespace": "openshift-sriov-network-operator", "name": "test-missing-ns", "reconcileID": "8267549b-ec6c-4370-a162-c4b88987c2aa", "sriovnetwork": "openshift-sriov-network-operator/test-missing-ns", "Namespace": "ns-xxx", "Name": "test-missing-ns", "error": "namespaces \"ns-xxx\" not found"}
2023-08-11T16:14:19.496675162Z  ERROR   Couldn't create NetworkAttachmentDefinition CR  {"controller": "sriovnetwork", "controllerGroup": "sriovnetwork.openshift.io", "controllerKind": "SriovNetwork", "SriovNetwork": {"name":"test-missing-ns","namespace":"openshift-sriov-network-operator"}, "namespace": "openshift-sriov-network-operator", "name": "test-missing-ns", "reconcileID": "7764a8ea-fd2e-4704-82b1-902f066e2cda", "sriovnetwork": "openshift-sriov-network-operator/test-missing-ns", "Namespace": "ns-xxx", "Name": "test-missing-ns", "error": "namespaces \"ns-xxx\" not found"}
2023-08-11T16:14:19.87228528Z   ERROR   Couldn't create NetworkAttachmentDefinition CR  {"controller": "sriovnetwork", "controllerGroup": "sriovnetwork.openshift.io", "controllerKind": "SriovNetwork", "SriovNetwork": {"name":"test-missing-ns","namespace":"openshift-sriov-network-operator"}, "namespace": "openshift-sriov-network-operator", "name": "test-missing-ns", "reconcileID": "5c111316-1313-470c-9e95-3637f5386bd1", "sriovnetwork": "openshift-sriov-network-operator/test-missing-ns", "Namespace": "ns-xxx", "Name": "test-missing-ns", "error": "namespaces \"ns-xxx\" not found"}
2023-08-11T16:14:20.569108499Z  ERROR   Couldn't create NetworkAttachmentDefinition CR  {"controller": "sriovnetwork", "controllerGroup": "sriovnetwork.openshift.io", "controllerKind": "SriovNetwork", "SriovNetwork": {"name":"test-missing-ns","namespace":"openshift-sriov-network-operator"}, "namespace": "openshift-sriov-network-operator", "name": "test-missing-ns", "reconcileID": "ac4bfd17-af0e-4421-8ec7-b583ce7c2831", "sriovnetwork": "openshift-sriov-network-operator/test-missing-ns", "Namespace": "ns-xxx", "Name": "test-missing-ns", "error": "namespaces \"ns-xxx\" not found"}
2023-08-11T16:14:21.905604308Z  ERROR   Couldn't create NetworkAttachmentDefinition CR  {"controller": "sriovnetwork", "controllerGroup": "sriovnetwork.openshift.io", "controllerKind": "SriovNetwork", "SriovNetwork": {"name":"test-missing-ns","namespace":"openshift-sriov-network-operator"}, "namespace": "openshift-sriov-network-operator", "name": "test-missing-ns", "reconcileID": "c63c700a-537d-4fc4-b64d-6b4dba8f7a4e", "sriovnetwork": "openshift-sriov-network-operator/test-missing-ns", "Namespace": "ns-xxx", "Name": "test-missing-ns", "error": "namespaces \"ns-xxx\" not found"}
```

### CODE REVIEW

File: controllers/sriovnetwork_controller.go [hunk 1]
```diff
@@ -141,6 +146,14 @@ func (r *SriovNetworkReconciler) Reconcile(ctx context.Context, req ctrl.Request
 	err = r.Get(ctx, types.NamespacedName{Name: netAttDef.Name, Namespace: netAttDef.Namespace}, found)
 	if err != nil {
 		if errors.IsNotFound(err) {
+			targetNamespace := &corev1.Namespace{}
+			err = r.Get(ctx, types.NamespacedName{Name: netAttDef.Namespace}, targetNamespace)
+			if errors.IsNotFound(err) {
+				// Target namespace doesn't exist, avoid retrying with exponential backoff
+				reqLogger.Info("Target namespace doesn't exist, cannot create NetworkAttachmentDefinition", "Namespace", netAttDef.Namespace, "Name", netAttDef.Name)
+				return reconcile.Result{}, nil
```
Comment (adrianchiris) [inline controllers/sriovnetwork_controller.go:154]: a less elegant solution is to just reconcile after X seconds using Result..RequeueAfter

this solution is much better !

File: controllers/sriovnetwork_controller.go [hunk 2]
```diff
@@ -21,13 +21,18 @@ import (
 	"reflect"
 
 	netattdefv1 "github.com/k8snetworkplumbingwg/network-attachment-definition-client/pkg/apis/k8s.cni.cncf.io/v1"
+	corev1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
```
Comment (adrianchiris) [inline controllers/sriovnetwork_controller.go:27]: nit: could you remove spaces here and below ?

File: controllers/sriovnetwork_controller.go [hunk 3]
```diff
@@ -141,6 +146,14 @@ func (r *SriovNetworkReconciler) Reconcile(ctx context.Context, req ctrl.Request
 	err = r.Get(ctx, types.NamespacedName{Name: netAttDef.Name, Namespace: netAttDef.Namespace}, found)
 	if err != nil {
 		if errors.IsNotFound(err) {
+			targetNamespace := &corev1.Namespace{}
+			err = r.Get(ctx, types.NamespacedName{Name: netAttDef.Namespace}, targetNamespace)
+			if errors.IsNotFound(err) {
+				// Target namespace doesn't exist, avoid retrying with exponential backoff
```
Comment (adrianchiris) [inline controllers/sriovnetwork_controller.go:152]: nit: perhaps reword this one 
> // Target namespace doesn't exist, avoid retrying with exponential backoff

// Target namespace doesn't exist. nothing to do. reconcile will be re-triggered when namespace is created.

or just remove comment and put info in log like:

reqLogger.Info("Target namespace doesn't exist, NetworkAttachmentDefinition will be created when namespace is available", "Namespace", netAttDef.Namespace, "Name", netAttDef.Name)

**Review by adrianchiris — COMMENTED** [review]
overall LGTM !

few small nits

@zeeke you mind adding the same logic for sriovIbNetwork controller ? i believe we have the same problem there as well.

if you dont have BW for it, can you create an issue to track ?

File: controllers/sriovnetwork_controller.go (hunk 3 above)
Comment (zeeke) [reply controllers/sriovnetwork_controller.go:152]: going for the latter

**Comment by zeeke** [comment]
> @zeeke you mind adding the same logic for sriovIbNetwork controller ? i believe we have the same problem there as well.

Done, please take a look

**Review by SchSeba — APPROVED** [review]
LGTM

nice work!

I learn about the index one it's nice!

**Review by adrianchiris — APPROVED** [review]
LGTM, just need to rebase this one and we are good 2 go 🚗

**Comment by adrianchiris** [comment]
> > @zeeke you mind adding the same logic for sriovIbNetwork controller ? i believe we have the same problem there as well.
> 
> Done, please take a look

Awesome Thanks !

File: controllers/sriovnetwork_controller.go [hunk 4]
```diff
@@ -173,8 +182,31 @@ func (r *SriovNetworkReconciler) Reconcile(ctx context.Context, req ctrl.Request
...
+			err := r.List(ctx,
+				&networkList,
+				client.MatchingFields{"spec.networkNamespace": e.Object.GetName()},
+			)
+			if err != nil {
+				log.Log.WithName("SriovNetworkReconciler").
+					Info("Can't list SriovNetworks for namespace", "resource", e.Object.GetName(), "error", err)
+			}
+
+			for _, network := range networkList.Items {
```
Comment (SchSeba) [inline controllers/sriovnetwork_controller.go:198]: just a small improvement can you check if the namespace in the handler is equal to the namespace of the network from the list?

Comment (zeeke) [reply controllers/sriovnetwork_controller.go:198]: Doesn't L191 ensure that? We are listing every SriovNetwork that has `spec.NetworkNamespace`  equal to the handler namespace.

Comment (SchSeba) [reply controllers/sriovnetwork_controller.go:198]: you are right!!
sorry I miss that one then all good

**Pull request merged**